### PR TITLE
feature test for share autocomplete list in webUI

### PR DESCRIPTION
**Pull request by haribhandari07**
## Description

Test for autocompletion feature in share dialog using webUI.
Includes test for allowing user to opt-out of autocomplete in sharing dialog #34942
and admin disabling share dialog user enumeration
## Related Issue

- Fixes <issue_link>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

File: tests/acceptance/features/lib/PersonalSharingSettingsPage.php [hunk 1]
```diff
@@ -43,7 +43,10 @@ class PersonalSharingSettingsPage extends SharingSettingsPage {
 		= '//label[@for="userAutoAcceptShareTrustedInput"]';
 	protected $autoAcceptFederatedSharesCheckboxXpathCheckboxId
 		= 'userAutoAcceptShareTrustedInput';
-
+	protected $allowFindingYouViaAutocompleteCheckboxXpath
+		= '//label[@for="allow_share_dialog_user_enumeration_input"]';
+	protected $allowFindingYouViaAutocompleteCheckboxXpathCheckboxId
```
Comment (dpakach) [inline tests/acceptance/features/lib/PersonalSharingSettingsPage.php]:
```suggestion
	protected $allowFindingYouViaAutocompleteCheckboxId
```

Comment (HariBhandari07) [reply tests/acceptance/features/lib/PersonalSharingSettingsPage.php]: I followed the legacy coding style: _
	protected $autoAcceptLocalSharesCheckboxXpath
	
	protected $autoAcceptLocalSharesCheckboxXpathCheckboxId
		
	protected $autoAcceptFederatedSharesCheckboxXpath
		
	protected $autoAcceptFederatedSharesCheckboxXpathCheckboxId

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 2]
```diff
@@ -895,6 +895,20 @@ public function theUsersOwnNameShouldNotBeListedInTheAutocompleteList() {
 			$this->sharingDialog->getAutocompleteItemsList()
 		);
 	}
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php]:
```suggestion
	}
	
```

File: tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature [hunk 3]
```diff
@@ -175,4 +175,68 @@ Feature: Autocompletion of share-with names
...
\ No newline at end of file
+    And user "User One" should not be listed in the autocomplete list on the webUI
+
+  @skipOnLDAP
+  Scenario: allow user to disable autocomplete in sharing dialog
+    Given user "regularuser" has logged in using the webUI
+    And the user has browsed to the personal sharing settings page
+    When the user disables allow finding you via autocomplete in share dialog
+    And the user logs out of the webUI
+    And user "user1" has logged in using the webUI
```
Comment (paurakhsharma) [inline tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature]:
```suggestion
    When the user re-logs in as "user1" using the webUI
```

We can use both logout and login step using one step.

File: tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature [hunk 4]
```diff
@@ -175,4 +175,68 @@ Feature: Autocompletion of share-with names
...
+    And the user types "reg" in the share-with-field
+    Then user "Regular User" should not be listed in the autocomplete list on the webUI
+
+  @skipOnLDAP
+  Scenario: allow user to enable autocomplete in sharing dialog
+    Given user "regularuser" has logged in using the webUI
+    And the user has browsed to the personal sharing settings page
+    When the user enables allow finding you via autocomplete in share dialog
+    And the user logs out of the webUI
+    And user "user1" has logged in using the webUI
```
Comment (paurakhsharma) [inline tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature]: Same here

File: tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature [hunk 5]
```diff
@@ -175,4 +175,68 @@ Feature: Autocompletion of share-with names
...
+    Then user "User One" should be listed in the autocomplete list on the webUI
+
+  @skipOnLDAP
+  Scenario: autocompletion of pattern when user disables and then enables autocompletion in sharing dialog
+    Given user "regularuser" has logged in using the webUI
+    And the user has browsed to the personal sharing settings page
+    When the user disables allow finding you via autocomplete in share dialog
+    And the user enables allow finding you via autocomplete in share dialog
+    And the user logs out of the webUI
+    And user "user1" has logged in using the webUI
```
Comment (paurakhsharma) [inline tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature]:
```suggestion
   When the user re-logs in as "user1" using the webUI
```

File: tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature (hunk 3 above)
Comment (HariBhandari07) [reply tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature]: Paurakh, here "regularuser" logs out and "user1" logs in, so logging out and logging in single step is not possible.

Comment (HariBhandari07) [reply tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature]: same as above

Comment (paurakhsharma) [reply tests/acceptance/features/webUISharingInternalUsers/shareAutocompletion.feature]: This step implementation doesn't take into account which user is currently logged in,
so this should be possible

File: tests/acceptance/features/bootstrap/WebUIPersonalSharingSettingsContext.php [hunk 6]
```diff
@@ -102,4 +102,28 @@ public function autoAcceptingFederatedCheckboxShouldNotBeDisplayedOnThePersonalS
 			$this->personalSharingSettingsPage->isAutoAcceptFederatedSharesCheckboxDisplayed()
 		);
 	}
+
+	/**
+	 * @Then allow finding you via autocomplete checkbox should not be displayed on the personal sharing settings page
+	 *
+	 * @return void
+	 */
+	public function allowFindingYouViaAutocompleteCheckboxShouldNotBeDisplayedOnThePersonalSharingSettingsPage() {
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIPersonalSharingSettingsContext.php]: Function name is way too long (must be auto-generated). You can truncate function name however you like. We try to cap line length to 79 chars (but is not a hard limit).

File: tests/acceptance/features/bootstrap/WebUIPersonalSharingSettingsContext.php [hunk 7]
```diff
@@ -102,4 +102,28 @@ public function autoAcceptingFederatedCheckboxShouldNotBeDisplayedOnThePersonalS
...
+		PHPUnit\Framework\Assert::assertFalse(
+			$this->personalSharingSettingsPage->isAllowFindingYouViaAutocompleteCheckboxDisplayed()
+		);
+	}
+
+	/**
+	 * @When /^the user (disables|enables) allow finding you via autocomplete in share dialog$/
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIPersonalSharingSettingsContext.php]: We try to order these kinds of step definitions based on behat keywords. So, `Given` steps are grouped at the top, then `When` steps and `Then` steps.

So, it'd be good to place this in the group of `When` step definitions.

File: tests/acceptance/features/bootstrap/WebUIPersonalSharingSettingsContext.php (hunk 6 above)
Comment (HariBhandari07) [reply tests/acceptance/features/bootstrap/WebUIPersonalSharingSettingsContext.php]: yup, it's long and was autogenerated but check out other functions name(for example in line 89, 100)

**Review by skshetry — COMMENTED** [review]
LGTM

**Review by paurakhsharma — DISMISSED** [review]
Looks good to me :+1:

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 8]
```diff
@@ -896,6 +896,21 @@ public function theUsersOwnNameShouldNotBeListedInTheAutocompleteList() {
...
+	 * @Then user :username should be listed in the autocomplete list on the webUI
+	 *
+	 * @param string $username
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function userShouldBeListedInTheAutocompleteListOnTheWebui($username) {
+		$names = $this->sharingDialog->getAutocompleteItemsList();
+		if (!\in_array($username, $names)) {
```
Comment (individual-it) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php]: use assert here

File: tests/acceptance/features/lib/PersonalSharingSettingsPage.php [hunk 9]
```diff
@@ -120,4 +138,17 @@ public function isAutoAcceptFederatedSharesCheckboxDisplayed() {
...
 		return true;
 	}
+
+	/**
+	 * Check if the allow finding you via autocomplete checkbox is shown in the webui
+	 *
+	 * @return boolean
+	 */
+	public function isAllowFindingYouViaAutocompleteCheckboxDisplayed() {
+		$localAutoAcceptCheckbox = $this->find('xpath', $this->allowFindingYouViaAutocompleteCheckboxXpath);
```
Comment (individual-it) [inline tests/acceptance/features/lib/PersonalSharingSettingsPage.php]: make line shorter

**Review by individual-it — CHANGES_REQUESTED** [review]
add tests 
- to check if user has disallowed autocompleting but sharer types full username
- to check if typing full display name when autocompletion is disabled (admin and user)

**Comment by HariBhandari07** [comment]
backport on #35217

**Review by individual-it — APPROVED** [review]

**Comment by individual-it** [comment]
@ownclouders rebase

**Pull request merged**